src/items.c

```diff
@@ -2831,3 +2859,249 @@ void HideSpawnPoints()
 	}
 }
 
+static void SetTiming(int item_class, float seconds, float lcl, float ucl)
+{
+	if (seconds == 0)
+	{
+		respawn_times[item_class] = default_respawn_times[deathmatch][item_class];
+	}
+	else if (default_respawn_times[deathmatch][item_class] < 0)
+	{
+		G_sprint(self, 2, "Error: item does not respawn in current deathmatch mode\n");
+	}
+	else if (seconds < lcl || seconds > ucl)
+	{
+		G_sprint(self, 2, "Error: duration out of range\n");
+	}
+	else
+	{
+		respawn_times[item_class] = seconds;
+	}
+}
+
+static void PrintTiming(int item_class, const char *prefix)
+{
+	float seconds;
+
+	seconds = respawn_times[item_class];
+	G_sprint(self, 2, "%s%s\n", prefix, seconds == -1 ? "never" : seconds == -2 ? "stay" : va("%0.3fs", seconds));
+}
+
+void ResetTimings()
+{
+	int i;
+
+	for (i = 0; i < NUM_ITEM_CLASSES; i++)
+	{
+		respawn_times[i] = default_respawn_times[deathmatch][i];
+	}
+}
+
+void TimingControl()
+{
+	int argc, allow;
+	float seconds;
+	char buf[64], *end;
+
+	argc = trap_CmdArgc();
+	if (argc < 2)
+	{
+		G_sprint(self, 2,
+			"Usage: timing %s|%s|%s|%s|%s|%s|%s|%s|%s [<%s>|%s]\n",
+			redtext("health"),
+			redtext("ammo"),
+			redtext("armor"),
+			redtext("weapon"),
+			redtext("mega"),
+			redtext("quad"),
+			redtext("eyes"),
+			redtext("pent"),
+			redtext("all"),
+			redtext("seconds"),
+			redtext("default")
+		);
+
+		return;
+	}
+
+	seconds = 0;
+	if (argc > 2)
+	{
+		allow = cvar("allow_custom_timings");
+		if (allow <= 0)
+		{
+			G_sprint(self, 2, "Error: custom timings not allowed on server\n");
+
+			return;
+		}
+
+		if (allow == 1)
+		{
+			if (!is_adm(self))
+			{
+				G_sprint(self, 2, "Error: administrator privileges required\n");
+
+				return;
+			}
+		}
+		else if (allow == 2)
+		{
+			if (self->ct != ctPlayer)
+			{
+				G_sprint(self, 2, "Error: spectators cannot change timings\n");
+
+				return;
+			}
+		}
+
+		if (!k_matchLess && match_in_progress)
+		{
+			G_sprint(self, 2, "Error: match currently in progress\n");
+
+			return;
+		}
+
+		trap_CmdArgv(2, buf, sizeof(buf));
+		if (!streq(buf, "default"))
+		{
+			seconds = strtof(buf, &end);
+			if (end == buf || seconds <= 0)
+			{
+				G_sprint(self, 2, "Error: invalid duration\n");
+
+				return;
+			}
+		}
+	}
+
+	trap_CmdArgv(1, buf, sizeof(buf));
+	if (streq(buf, "health"))
+	{
```

Comment:
Let's introduce a helper function here. Something like 

```
if (streq(buf, "health"))
	{
		PrintSetTiming("health", ITEM_CLASS_HEALTH, seconds, argc == 2);
	}
	else if (streq(buf, "ammo"))
	{
		PrintSetTiming("ammo", ITEM_CLASS_AMMO, seconds, argc == 2);
	}
	else if (streq(buf, "armor"))
	{
		PrintSetTiming("armor", ITEM_CLASS_ARMOR, seconds, argc == 2);
	}
...

PrintSetTiming(char *item, int itemClass, float seconds, qbool print)
{
	if (print)
	{
		char prefix[20] = "";
		sprintf(prefix, "%s respawn time: ", *item);
		PrintTiming(itemClass, prefix);
	}
	else
	{
		SetTiming(itemClass, seconds, ITEM_TIME_MIN, ITEM_TIME_MAX);
	}
}
```

Reply:
I guess so? What will you do in the case of "all"?

Reply:
We are repeating the same 'code' many times.
Even if I do the same thing only twice, I consider a helper function. But in this case, when we repeat it like 7 times, so I think a helper function will make this a much cleaner and compact piece of code, won't it?

Reply:
You're right that it's kind of repeating, and I can see how that might seem sub-optimal, but it's not _exactly_ repeating. _PrintTiming_ is already a helper function for this very reason, and it's being called with different arguments every time.

I like these little helper functions to do one simple thing. _PrintSetTiming_ changes state in some cases and leaves state alone in others. I'm not sure that is actually cleaner than what exists currently.